### DISCUSSION TITLE
Eliminating warnings

### DISCUSSION THREAD

**Discussion by bobwolff68** [Bodmer/TFT_eSPI#1109] [General]
Hi there,
I'm a new user to your library - switching over from the Adafruit setup. Very nice library, btw. Super impressed by its depth and breadth.

I'm using an ESP32 + ILI9341 display building from PlatformIO.

I do try to keep warnings out of my compiles so that when a warning does come along I see it and can decide if it is significant. I've been bitten by "too much noise" in my past - having warnings always present in my compilations causing me to not see a new one which might be important. :)

So, I'm seeing warnings about re-definitions. It's a byproduct of using the compiler's "-D" to define things like TFT_WIDTH 

```Compiling .pio/build/esp32thing/src/ILI9341_UnitTest.cpp.o
In file included from .pio/libdeps/esp32thing/TFT_eSPI/User_Setup_Select.h:133:0,
                 from .pio/libdeps/esp32thing/TFT_eSPI/TFT_eSPI.h:39,
                 from /Users/rwolff/dev/ArduinoProjects/roboswamp/lib/robolib/TFT_TempGauge.h:6,
                 from /Users/rwolff/dev/ArduinoProjects/roboswamp/lib/robolib/RoboSwamp.h:44,
                 from src/ILI9341_UnitTest.cpp:18:
.pio/libdeps/esp32thing/TFT_eSPI/TFT_Drivers/ILI9341_Defines.h:3:0: warning: "TFT_WIDTH" redefined
 #define TFT_WIDTH  240
 ^
<command-line>:0:0: note: this is the location of the previous definition
```
The offending source file in my case is ILI9341_Defines.h
```// Change the width and height if required (defined in portrait mode)
// or use the constructor to over-ride defaults
#define TFT_WIDTH  240
#define TFT_HEIGHT 320
```

If using the compiler '-D' is the preferred/supported method of doing things, might I suggest doing an `#ifndef` around these redefinitions?
```// Change the width and height if required (defined in portrait mode)
// or use the constructor to over-ride defaults
#ifndef TFT_WIDTH
#define TFT_WIDTH  240
#endif
#ifndef TFT_HEIGHT
#define TFT_HEIGHT 320
#endif
```

I'd be happy to give a pull request on the ones I find if that's helpful. I understand to do a 'complete' job at this it would impact other header files for hardware I'm not using etc. Let me know!

**Comment by Bodmer** [comment]
There was a typo in the Setup.h from a recent update that I have corrected.

Some displays have the same driver but different widths and heights are available. The width and height only needs to be specified for those types.[ See here](https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setup.h#L77-L86). For these displays the #ifdef approach proposed is used, however for consistency this approach could have been used for all display types but this can lead to problematic "user" defines because the library expects the width and height to be defined portrait mode.

For other displays driver types the width and height defaults should be fine so these do not need to be duplicated in the setup configuration.

**Reply by bobwolff68** [reply]
Yeah I was afraid the complexity of what you're offering here (many options/drivers/etc) might complicate an otherwise nice item for removing a few warnings.

I'm just reading this notion about the library expecting the width and height to be defined in portrait mode. I had noticed that and when I used TFT_WIDTH=320 and TFT_HEIGHT=240, I wound up with some odd graphics from prior runs that didn't get cleared after a reset and wound up swapping the two resolutions to fix it. I see the mention of it in the header file now.
